runtime/src/iree/hal/drivers/hip/hip_device_group_device.c

```diff
@@ -0,0 +1,1535 @@
+    uint32_t device_count, hipDevice_t* devices,
+    iree_allocator_t host_allocator, iree_hal_device_t** out_device) {
+  IREE_ASSERT_ARGUMENT(driver);
+  IREE_ASSERT_ARGUMENT(params);
+  IREE_ASSERT_ARGUMENT(symbols);
+  IREE_ASSERT_ARGUMENT(out_device);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_hip_device_group_device_t* device;
+  const iree_host_size_t total_device_size =
+      sizeof(*device) +
+      sizeof(iree_hal_hip_per_device_information_t) * device_count +
+      identifier.size;
+  IREE_RETURN_IF_ERROR(iree_allocator_malloc(host_allocator, total_device_size,
+                                             (void**)&device));
+  device->num_physical_devices = device_count;
+
+  iree_status_t status = iree_hal_hip_device_group_device_check_params(params);
+
+  // Get the main context for the device.
+  for (uint32_t i = 0; i < device_count && iree_status_is_ok(status); ++i) {
+    device->device_contexts[i].hip_device = devices[i];
+    status = IREE_HIP_RESULT_TO_STATUS(
+        symbols, hipDevicePrimaryCtxRetain(
+                     &device->device_contexts[i].hip_context, devices[i]));
+    if (iree_status_is_ok(status)) {
+      status = IREE_HIP_RESULT_TO_STATUS(
+          symbols, hipCtxSetCurrent(device->device_contexts[i].hip_context));
+    }
+
+    // Create the default dispatch stream for the device.
+    if (iree_status_is_ok(status)) {
+      status = IREE_HIP_RESULT_TO_STATUS(
+          symbols, hipStreamCreateWithFlags(
+                       &device->device_contexts[i].hip_dispatch_stream,
+                       hipStreamNonBlocking));
+    }
+
+    if (iree_status_is_ok(status)) {
+      for (uint32_t j = 0; j < device_count && iree_status_is_ok(status); ++j) {
+        if (i == j) {
+          continue;
+        }
+        status = IREE_HIP_RESULT_TO_STATUS(
+            symbols, hipDeviceEnablePeerAccess(devices[j], 0));
+      }
+    }
+  }
+
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_hip_device_group_device_initialize_internal(
+        driver, identifier, params, device, symbols, nccl_symbols,
+        host_allocator);
+  } else {
+    for (uint32_t i = 0; i < device_count && iree_status_is_ok(status); ++i) {
+      if (device->device_contexts[i].hip_dispatch_stream)
+        symbols->hipStreamDestroy(
+            device->device_contexts[i].hip_dispatch_stream);
+      // NOTE: This function return hipSuccess though doesn't release the
+      // primaryCtx by design on HIP/HCC path.
+      if (device->device_contexts[i].hip_context)
+        symbols->hipDevicePrimaryCtxRelease(
+            device->device_contexts[i].hip_device);
+    }
+  }
+
+  iree_event_pool_t* host_event_pool = NULL;
+  if (iree_status_is_ok(status)) {
+    status = iree_event_pool_allocate(params->event_pool_capacity,
+                                      host_allocator, &host_event_pool);
+  }
+
+  for (uint32_t i = 0; i < device_count && iree_status_is_ok(status); ++i) {
```

Comment:
style: prefer `iree_host_size_t` for count/iterator unless optimizing for at-rest in-memory size (when passing args and stuff we aren't - or will let the compiler do it if it wants).